refinery/ui/source/js/data-set-import/controllers/file-upload-s3.js

```diff
@@ -5,25 +5,133 @@
-          }, function (progress) {
-            // Write the progress as a percentage
+    vm.multifileUploadInProgress = false;
+
+    vm.isUploadConfigured = function () {
+      if (s3UploadService.isConfigReady) {
+        return s3UploadService.isConfigValid();
+      }
+      return false;
+    };
+
+    vm.addFiles = function (files) {
+      if (typeof vm.files === 'undefined') {
+        vm.files = files;
+      } else {
+        vm.files = vm.files.concat(files);
+      }
+    };
+
+    vm.isFileNew = function (file) {
+      // check if file upload has been attempted
+      return typeof file.progress === 'undefined';
+    };
+
+    vm.isUploadInProgress = function (file) {
+      if (vm.isFileNew(file) || file.$error || file.success) {
+        return false;
+      }
+      return file.progress <= 100;
+    };
+
+    vm.areUploadsInProgress = function () {
+      if (typeof vm.files === 'undefined') {
+        return false;
+      }
+      return vm.files.some(vm.isUploadInProgress);
+    };
+
+    vm.isUploadComplete = function (file) {
+      return Boolean(file.$error || file.success);
+    };
+
+    vm.areUploadsCancellable = function () {
+      if (typeof vm.files === 'undefined') {
+        return false;
+      }
+      return !vm.files.every(vm.isUploadComplete);
+    };
+
+    vm.areUploadsEnabled = function () {
+      if (typeof vm.files === 'undefined' || vm.areUploadsInProgress()) {
+        return false;
+      }
+      return vm.files.some(vm.isFileNew);
+    };
+
+    vm.uploadFile = function (file) {
+      try {
+        file.managedUpload = s3UploadService.upload(file);
+      } catch (e) {
+        file.progress = 100;
+        file.$error = 'Data upload configuration error';
+        return;
+      }
+      file.progress = 0;
+      file.managedUpload.on('httpUploadProgress', function (progress) {
+        // $applyAsync is used to avoid $rootScope:inprog error when canceling uploads
+        $scope.$applyAsync(function () {
+          if (progress.total) {
             file.progress = (progress.loaded / progress.total) * 100;
-          });
+          }
+        });
+      });
+      file.managedUpload.promise().then(function () {
+        $scope.$apply(function () {
+          file.success = true;
+          if (vm.multifileUploadInProgress) {
+            vm.uploadFiles();
+          }
         });
+      }, function (error) {
+        $scope.$apply(function () {
+          file.progress = 100;
+          file.$error = error;
+          $log.error('Error uploading file ' + file.name + ': ' + file.$error);
+          if (vm.multifileUploadInProgress) {
+            vm.uploadFiles();
+          }
+        });
+      });
+    };
+
+    vm.uploadFiles = function () {
+      if (vm.files) {
+        vm.multifileUploadInProgress = true;
+        for (var i = 0; i < vm.files.length; i++) {
+          if (vm.isFileNew(vm.files[i])) {
+            vm.uploadFile(vm.files[i]);
+            return;
```

Comment:
Exactly: `uploadFiles()` is called from `uploadFile()`. It is a bit too clever, so I'll add a comment.